# DTSC-580: Data Manipulation
## Assignment: Scooby-Doo

### Name: Enter Name Before Submitting

## Copyright & Academic Integrity Notice
<span style="color:red">This material is for enrolled students' academic use only and protected under U.S. Copyright Laws. This content must not be shared outside the confines of this course, in line with Eastern University's academic integrity policies. Unauthorized reproduction, distribution, or transmission of this material, including but not limited to posting on third-party platforms like GitHub, is strictly prohibited and may lead to disciplinary action. You may not alter or remove any copyright or other notice from copies of any content taken from BrightSpace or Eastern University’s website.</span>

<span style="color:red">© Copyright Notice 2025, Eastern University - All Rights Reserved.</span>

## Overview

In this assignment, you will test the skills that you have learned during this course to manipulate the provided data to answer questions about the TV show Scooby-Doo.  If you are not familiar with this show, more information can be found on [Scoobypedia](https://scoobydoo.fandom.com/wiki/Scoobypedia).

> Scoobypedia is an encyclopedia on the hit television series Scooby-Doo which has been airing for over 50 years!
>
> The show follows the iconic mystery solving detectives, known as Mystery Inc., as they set out in the Mystery Machine to solve crime and unmask criminals, bent on revenge or committing criminal acts for their own personal gain.
>
> Titular character, Scooby, is accompanied by his best pal Shaggy as both vie for Scooby Snacks on their adventures! Velma brings her laser-focused intellect and initiative to collecting clues and solving mysteries. Fred is the team's leader who devises elaborate traps to catch the bad guys, while Daphne is bold and empathetic and keeps the Gang together.

Please note that your notebook should be named `scooby` when submitting to CodeGrade for the automatic grading to work properly.

## About Dataset

This data comes from [Kaggle](https://www.kaggle.com/datasets/williamschooleman/scoobydoo-complete) thanks to manual data aggregation by Kaggle user plummye. We have included a PDF from the dataset's author that provides detailed information on how the data was collected, additional details about the features, and how the data was analyzed to identify features potentially important to the IMDB score.

## Data Dictionary

### scoobydoo_episodes
| variable | description |
| ----------| ---------- |
| index	| index based on Scoobypedia |
| series_name | Name of the series in which the episode takes place or in movies' cases the Scoobypedia's grouping classification |
| network | Network the TV series takes place in, if it is a movie will use similar grouping as series.name variable |
| season | Season of TV Series, if not TV Series will default to the format |
| title | Title of Episode/Movie |
| imdb | Score on IMDB (NULL if recently aired) |
| engagement | Number of reviews on IMDB (NULL if very recently aired) |
| date_aired | Dated aired in US |
| run_time | Run time in min |
| format | Type |

### scoobydoo_monsters
| variable | description |
| ---------- | ---------|
|index | index based on Scoobypedia |
|monster_name | name of monster |
|monster_gender | monster gender |
|monster_type | monster type |
|monster_subtype | monster subtype |
|monster_species | monster_species |
|monster_real | was monster real |
|monster_amount | monster amount |
|caught_fred | caught by Fred |
|caught_daphnie | caught by Daphnie |
|caught_velma | caught by Velma |
|caught_shaggy | caught by Shaggy |
|caught_scooby | caught by Scooby |
|captured_fred | captured Fred |
|captured_daphnie |	captured Daphnie |
|captured_velma  | captured Velma |
|captured_shaggy | captured Shaggy |
|captured_scooby | captured Scooby |
|unmask_fred | unmask by fred |
|unmask_daphnie | unmask by Daphnie |
|unmask_velma | unmask by Velma |
|unmask_shaggy | unmask by Shaggy |
|unmask_scooby | unmask by Scooby |
|snack_fred | snack offered by Fred |
|snack_daphnie | snack offered by Daphnie |
|snack_velma | snack offered by Velma |
|snack_shaggy | snack offered by Shaggy |
|snack_scooby | snack offered by Scooby |
|unmask_other |	unmask by other |
|caught_other | caught by other |
|caught_not | not caught |
|trap_work_first | trap work first |
|setting_terrain | setting type of terrain |
|setting_country_state | setting country state |
|suspects_amount | suspects amount |
|non_suspect | non suspect |
|arrested | arrested |
|culprit_name |	culprit name |
|culprit_gender | culprit gender |
|culprit_amount | culprit amount |
|motive | motive |
|if_it_wasnt_for | phrase at the end of show, ie "if it wasnt for ..." |
|and_that | and that |
|door_gag | door gag |
|number_of_snacks | number of snacks |
|split_up |	split up |
|another_mystery | another mystery |
|set_a_trap | set a trap |
|jeepers | Times "jeepers" said |
|jinkies | Times "jinkies" said |
|my_glasses	| Times "my glasses" said |
|just_about_wrapped_up | Times "just about wrapped up" said |
|zoinks	| Times "zoinks"said |
|groovy	| Times "groovy" said |
|scooby_doo_where_are_you | Times "scooby doo where are you" said |
|rooby_rooby_roo | Times "rooby_rooby_roo" said |
|batman | batman in episode |
|scooby_dum | scooby_dum in episode |
|scrappy_doo | scrappy_doo in episode |
|hex_girls | hex_girls in episode |
|blue_falcon | blue_falcon in episode |
|fred_va | Fred voice actor |
|daphnie_va | Daphnie voice actor |
|velma_va | Velma voice actor |
|shaggy_va | Shaggy voice actor |
|scooby_va | Scooby voice actor |

## Note

<u>Show Work</u>

Remember that you must show your work.  Students submissions are spot checked manually to verify that they are not hard coding the answer from looking only in the file or in CodeGrade's expected output.  If this is seen, the student's answer will be manually marked wrong and their grade will be changed to reflect this.


## Initial Import & Exploration

In [1]:
# standard imports
import pandas as pd
import numpy as np

# Do not change this option; This allows the CodeGrade auto grading to function correctly
pd.set_option('display.max_columns', None)
import warnings
warnings.filterwarnings("ignore")

Start by importing your two data files. Take time to thoroughly explore the data. Investing effort in understanding the dataset upfront typically makes data manipulation tasks more efficient and manageable during a project.

In [2]:
### Import Data Files ###
scoobydoo_episodes = pd.read_csv('scoobydoo_episodes.csv')
scoobydoo_monsters = pd.read_csv('scoobydoo_monsters.csv')

In [3]:
### Explore Data ###
scoobydoo_episodes.info()
scoobydoo_monsters.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 608 entries, 0 to 607
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Index        608 non-null    int64  
 1   Series Name  608 non-null    object 
 2   Network      608 non-null    object 
 3   Season       608 non-null    object 
 4   Title        608 non-null    object 
 5   IMDB         593 non-null    float64
 6   Engagement   589 non-null    float64
 7   Date Aired   608 non-null    object 
 8   Run Time     590 non-null    object 
 9   Format       608 non-null    object 
dtypes: float64(2), int64(1), object(7)
memory usage: 47.6+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 608 entries, 0 to 607
Data columns (total 66 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Index                     608 non-null    int64  
 1   Monster Name              520 non-null    object 
 2 

## Exercises

**Exercise1: Merging Datasets**

Merge the two datasets and name the resulting DataFrame `scoobydoo_merged`.

In [4]:
### ENTER CODE HERE ###
scoobydoo_merged = pd.merge(scoobydoo_episodes, scoobydoo_monsters, on='Index')

**Code Check:** The shape of `scoobydoo_merged` should be `(608, 75)` after this initial merge. Note that the shape of the dataset will be different  at the end of the assignment due to some cleaning steps that follow.

In [5]:
### CODE CHECK ###
scoobydoo_merged.shape
#scoobydoo_merged.info()

(608, 75)

<hr/>

**Exercise2: Updating Column Names**

Change every column name to lowercase and snake case.  This is a standard first step for some programmers as lowercase makes it easier to write and snake case makes it easier to copy multiple-word column names.

For example, `Series Name` should end up being `series_name`. It's fine if you don't, but you should be able to accomplish this in a single line of code without using a for loop.

In [6]:
### ENTER CODE HERE ###
scoobydoo_merged.columns = scoobydoo_merged.columns.str.lower().str.replace(' ', '_')

**Code Check:** Check your columns to make sure your changes worked as expected and were saved to the `scoobydoo_merged` DataFrame.

In [7]:
### CODE CHECK ###
scoobydoo_merged.columns

Index(['index', 'series_name', 'network', 'season', 'title', 'imdb',
       'engagement', 'date_aired', 'run_time', 'format', 'monster_name',
       'monster_gender', 'monster_type', 'monster_subtype', 'monster_species',
       'monster_real', 'monster_amount', 'caught_fred', 'caught_daphnie',
       'caught_velma', 'caught_shaggy', 'caught_scooby', 'captured_fred',
       'captured_daphnie', 'captured_velma', 'captured_shaggy',
       'captured_scooby', 'unmask_fred', 'unmask_daphnie', 'unmask_velma',
       'unmask_shaggy', 'unmask_scooby', 'snack_fred', 'snack_daphnie',
       'snack_velma', 'snack_shaggy', 'snack_scooby', 'unmask_other',
       'caught_other', 'caught_not', 'trap_work_first', 'setting_terrain',
       'setting_country_state', 'suspects_amount', 'non_suspect', 'arrested',
       'culprit_name', 'culprit_gender', 'culprit_amount', 'motive',
       'if_it_wasnt_for', 'and_that', 'door_gag', 'number_of_snacks',
       'split_up', 'another_mystery', 'set_a_trap', 'jeepe

<hr/>

**Exercise3: Handle Duplicates**

1) Identify and remove any duplicate rows saving your changes back to `scoobydoo_merged`. For this assignment, assume duplicates have the same `title` and `date_aired`.
2) Reset the index for `scoobydoo_merged` ensuring that it goes from 0 to n-1.

In [8]:
### ENTER CODE HERE ###
scoobydoo_merged.drop_duplicates(subset=['title', 'date_aired'], inplace=True)
scoobydoo_merged.reset_index(drop=True, inplace=True)

**Code Check:** Check the shape of `scoobydoo_merged`; it should now be `(603, 75)`.

In [9]:
### CODE CHECK ###
scoobydoo_merged.shape

(603, 75)

<hr/>

**Exercise4: Clean Runtime Column Part 1**

To get ready for a later exercise, we need to clean the `run_time` column. Let's start by updating all `run_time` values in `scoobydoo_merged` to floats (e.g., clean values like `21 mins` to the float `21.0`).

In [10]:
### ENTER CODE HERE ###
scoobydoo_merged['run_time'] = scoobydoo_merged['run_time'].str.replace(' mins', '').astype(float)

**Code Check:** Check that the first five values of the `run_time` column match the following output:

```
0      21.0
1      22.0
2      21.0
3      21.0
4      21.0
```

In [11]:
### CODE CHECK ###
scoobydoo_merged['run_time'].head()
scoobydoo_merged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 603 entries, 0 to 602
Data columns (total 75 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   index                     603 non-null    int64  
 1   series_name               603 non-null    object 
 2   network                   603 non-null    object 
 3   season                    603 non-null    object 
 4   title                     603 non-null    object 
 5   imdb                      588 non-null    float64
 6   engagement                584 non-null    float64
 7   date_aired                603 non-null    object 
 8   run_time                  585 non-null    float64
 9   format                    603 non-null    object 
 10  monster_name              516 non-null    object 
 11  monster_gender            516 non-null    object 
 12  monster_type              516 non-null    object 
 13  monster_subtype           515 non-null    object 
 14  monster_sp

<hr/>

**Exercise5: Clean Runtime Column Part 2**

The `run_time` column has missing values for each format type. Impute these missing values by replacing them with the average run time for their respective format type rounded to two decimals. For example, use the average run time for the "TV Series" format (rounded to 2 decimals) to fill null values for rows with that format, use the average run time for the "TV Series (segmented)" format (rounded to 2 decimals) to fill null values for rows with that format, etc.

In [12]:
### ENTER CODE HERE ###
mean_run_time = scoobydoo_merged.groupby('format')['run_time'].transform('mean').round(2)
scoobydoo_merged['run_time'].fillna(mean_run_time, inplace=True)
scoobydoo_merged.shape

(603, 75)

**Code Check:** 1) Ensure the `run_time` column has no missing values. 2) To confirm you imputed the correct mean values, the sum of the column should now be `14183.73`.

In [13]:
### CODE CHECK ###
scoobydoo_merged['run_time'].isna().sum()
scoobydoo_merged['run_time'].sum()

np.float64(14183.73)

<hr/>

**Exercise6: Summarize Runtime by Format**

With the `run_time` column cleaned, analyze how the average run time differs across the various format types. Group your data to determine the average run time for each format type, rounded to two decimal places. Arrange your results in ascending order and save them as a Series named `Exercise6`.

In [14]:
### ENTER CODE HERE ###
Exercise6 = scoobydoo_merged.groupby('format')['run_time'].mean().round(2).sort_values()
Exercise6.head()
scoobydoo_merged

index                 series_name            network season  \
0        1  Scooby Doo, Where Are You!                CBS      1   
1        2  Scooby Doo, Where Are You!                CBS      1   
2        3  Scooby Doo, Where Are You!                CBS      1   
3        4  Scooby Doo, Where Are You!                CBS      1   
4        5  Scooby Doo, Where Are You!                CBS      1   
..     ...                         ...                ...    ...   
598    604   Scooby-Doo and Guess Who?          Boomerang      2   
599    605           Warner Home Video  Warner Home Video  Movie   
600    606           Warner Home Video  Warner Home Video  Movie   
601    607   Scooby-Doo and Guess Who?          Boomerang      2   
602    608   Scooby-Doo and Guess Who?          Boomerang      2   

                                    title  imdb  engagement  date_aired  \
0               What a Night for a Knight   8.1       556.0   9/13/1969   
1                   A Clue for Scooby Doo   8.1       479.0   9/20/1969   
2                    Hassle in the Castle   8.0       455.0   9/27/1969   
3                  Mine Your Own Business   7.8       426.0   10/4/1969   
4                   Decoy for a Dognapper   7.5       391.0  10/11/1969   
..                                    ...   ...         ...         ...   
598  The Dreaded Remake of Jekyll & Hyde!   NaN         NaN   10/1/2020   
599          Happy Halloween, Scooby-Doo!   6.3       978.0   10/6/2020   
600   Scooby-Doo! The Sword and the Scoob   6.2       422.0   2/23/2020   
601               Dark Diner of Route 66!   NaN         NaN   2/25/2021   
602                       Total Jeopardy!   NaN         NaN  11/13/2020   

     run_time     format                                 monster_name  \
0        21.0  TV Series                                 Black Knight   
1        22.0  TV Series                       Ghost of Cptn. Cuttler   
2        21.0  TV Series                                      Phantom   
3        21.0  TV Series                                  Miner 49'er   
4        21.0  TV Series                          Indian Witch Doctor   
..        ...        ...                                          ...   
598      22.0  TV Series  Hyde Monster,Mummy,Dinosaur,Giant Manta Ray   
599      76.0      Movie         Alpha Jackal-Lantern,Jackal-Lanterns   
600      77.0      Movie                   Morgan le Fay,Black Knight   
601      22.0  TV Series                      Mud Men,Mud Men,Mud Men   
602      23.0  TV Series                                 Max Kilobyte   

          monster_gender                                       monster_type  \
0                   Male                                   Possessed Object   
1                   Male                                              Ghost   
2                   Male                                              Ghost   
3                   Male                                            Ancient   
4                 Female                                            Ancient   
..                   ...                                                ...   
598  Male,Male,Male,Male                     Disguised,Undead,Animal,Animal   
599            Male,Male                                        Plant,Plant   
600          Female,Male                          Magician,Possessed Object   
601       Male,Male,Male  Possessed Object,Possessed Object,Possessed Ob...   
602                 Male                                         Mechanical   

                            monster_subtype                 monster_species  \
0                                      Suit                          Object   
1                                      Suit                           Human   
2                                   Phantom                           Human   
3                                     Miner                           Human   
4                              Witch Doctor                       

<hr/>

**Exercise7: Clean the Network Column**

Address the inconsistencies in the network column by standardizing entries for 'Cartoon Network.' Ensure all variations of the name (e.g., spelling differences) are replaced with a consistent format: 'Cartoon Network.'"

Hint: You can reuse code from a previous assignment for this step. As your data manipulation skills grow, you'll notice recurring patterns in your tasks. Building a library of reusable code will be valuable for future projects.

In [15]:
### ENTER CODE HERE ###
scoobydoo_merged['network'] = scoobydoo_merged['network'].str.replace('.*cartoon.*', 'Cartoon Network', case=False, regex=True)


**Code Check:** Display the unique network values to verify that all variations of "Cartoon Network" have been combined.

In [16]:
### CODE CHECK ###
scoobydoo_merged['network'].unique()
scoobydoo_merged.shape

(603, 75)

<hr/>

**Exercise8: Network with Highest IMDB Scores**

With the `network` column cleaned, analyze the average IMDB score for each network. Identify the network with the highest average score by grouping the data, calculating the averages (rounded to two decimals), and organizing the results in ascending order. Save your final output as a Series named `Exercise8`.

In [17]:
### ENTER CODE HERE ###
Exercise8 = scoobydoo_merged.groupby('network')['imdb'].mean().round(2).sort_values()
Exercise8.head()

network
Warner Bros. Picture    5.30
The CW                  5.75
TBC                     5.80
Warner Home Video       6.58
Syndication             6.97
Name: imdb, dtype: float64

<hr/>

**Exercise9: Who Caught the Most Monsters**

Determine which characters caught the most monsters: Fred, Daphne, Velma, Shaggy, or Scooby. Use the columns related to monster captures (refer to the data dictionary for details) to calculate the total number of captures for each character. Sort the results in ascending order and save your final output as a Series named `Exercise9`.

In [18]:
### ENTER CODE HERE ###
monster_columns = ['caught_fred', 'caught_daphnie', 'caught_velma', 'caught_shaggy', 'caught_scooby']
scoobydoo_monster = scoobydoo_merged.copy()
scoobydoo_monster['total_monsters_caught'] = scoobydoo_monster[monster_columns].sum(axis=1)
Exercise9 = scoobydoo_monster[monster_columns].sum().sort_values(ascending=True)
Exercise9
scoobydoo_merged.shape

(603, 75)

<hr/>

**Exercise10: Who Was Most Succcessful At Traps**

In the previous exercise, you identified who caught the most monsters. Now, refine this analysis by focusing only on episodes where the trap worked on the first try. Determine who was the most successful in catching monsters under these conditions.

Sort the results in ascending order and save the result as a Series named `Exercise10`.

Hint: You may find the code from the previous exercise helpful but will need to add an additional condition.

In [19]:
### ENTER CODE HERE ###

scoobydoo_monster['total_monsters_caught_trap_first'] = scoobydoo_monster[monster_columns].sum(axis=1)
Exercise10 = scoobydoo_monster[scoobydoo_monster['trap_work_first'] == 1][monster_columns].sum().sort_values(ascending=True)
Exercise10
scoobydoo_merged.shape

(603, 75)

<hr/>

**Exercise11: Add Year Aired Column**

You want to determine when the show was at its peak based on total number of episodes per year. In order to get this information, you'll add a new column called `year_aired` that strips the year from the `date_aired` column.

1) Make a copy of the `scoobydoo_merged` DataFrame and name it `scoobydoo_merged_year`. This ensures the original data remains unchanged, making CodeGrade grading easier.
2) Add a new column, `year_aired`, to the copied DataFrame. Extract the year from the `date_aired` column and assign it to this new column. Ensure that the new column is an integer data type.

In [21]:
### ENTER CODE HERE ###]
scoobydoo_merged_year = scoobydoo_merged.copy()
scoobydoo_merged_year['year_aired'] = pd.to_datetime(scoobydoo_merged_year['date_aired']).dt.year.astype(int)
scoobydoo_merged_year.info()
scoobydoo_merged.shape

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 603 entries, 0 to 602
Data columns (total 76 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   index                     603 non-null    int64  
 1   series_name               603 non-null    object 
 2   network                   603 non-null    object 
 3   season                    603 non-null    object 
 4   title                     603 non-null    object 
 5   imdb                      588 non-null    float64
 6   engagement                584 non-null    float64
 7   date_aired                603 non-null    object 
 8   run_time                  603 non-null    float64
 9   format                    603 non-null    object 
 10  monster_name              516 non-null    object 
 11  monster_gender            516 non-null    object 
 12  monster_type              516 non-null    object 
 13  monster_subtype           515 non-null    object 
 14  monster_sp

(603, 75)

**Code Check:** Select the `date_aired` and `year_aired` columns from the `scoobydoo_merged_year` DataFrame. The output should match the following for the first and last five rows:

```
date_aired  year_aired
0     9/13/1969        1969
1     9/20/1969        1969
2     9/27/1969        1969
3     10/4/1969        1969
4    10/11/1969        1969
..          ...         ...
598   10/1/2020        2020
599   10/6/2020        2020
600   2/23/2020        2020
601   2/25/2021        2021
602  11/13/2020        2020
```

In [22]:
### CODE CHECK ###
scoobydoo_merged_year[['date_aired', 'year_aired']]

date_aired  year_aired
0     9/13/1969        1969
1     9/20/1969        1969
2     9/27/1969        1969
3     10/4/1969        1969
4    10/11/1969        1969
..          ...         ...
598   10/1/2020        2020
599   10/6/2020        2020
600   2/23/2020        2020
601   2/25/2021        2021
602  11/13/2020        2020

[603 rows x 2 columns]

<hr/>

**Exercise12: Total Number of Episodes Per Year**

Use the `scoobydoo_merged_year` DataFrame to create a Series showing the total number of episodes per year. Ensure the results are in **ascending** year order and save the result as a Series named `Exercise12`.

In [24]:
### ENTER CODE HERE ###
Exercise12 = scoobydoo_merged_year.groupby('year_aired').size().sort_index(ascending=True)
Exercise12
scoobydoo_merged.shape
scoobydoo_merged_year.shape

(603, 76)

<hr/>

**Exercise13: Total Number of Episodes Per Year and Per Format**

Using the `scoobydoo_merged_year` DataFrame, explore how the total number of episodes varies across years and formats. Organize the data to calculate the total episodes for each unique combination of year and then format. Keep the results in their original order, and assign your output to a Series named `Exercise13`.

In [25]:
### ENTER CODE HERE ###
Exercise13 = scoobydoo_merged_year.groupby(['year_aired', 'format']).size()
Exercise13

year_aired  format            
1969        TV Series             15
1970        TV Series             10
1972        TV Series             16
1973        TV Series              8
1976        Crossover              3
                                  ..
2019        TV Series             13
2020        Movie                  2
            Movie (Theatrical)     1
            TV Series             27
2021        TV Series              1
Length: 80, dtype: int64

<hr/>

**Exercise14: Who Offered the Most Snacks**

According to the creator of the dataset, the `snack_fred`, `snack_daphnie`, `snack_velma`, `snack_shaggy`, and `snack_scooby` columns refers to "when a member of the gang offers someone a Scooby Snack (usually
given to Shaggy and Scooby). The person they offer does not have to eat it for this to count as TRUE"

Who offers the most snacks? Going back to the `scoobydoo_merged` DataFrame, create a Series that shows the total number of times a scooby snack was offered by each character. Sort the values in descending order, and save the output as a Series named `Exercise14`. You will only include the five columns mentioned above.

In [26]:
### ENTER CODE HERE ###
snack_columns = ['snack_fred', 'snack_daphnie', 'snack_velma', 'snack_shaggy', 'snack_scooby']
Exercise14 = scoobydoo_merged[snack_columns].sum().sort_values(ascending=False)
Exercise14

snack_daphnie    49
snack_shaggy     43
snack_velma      29
snack_fred       18
snack_scooby     12
dtype: object

<hr/>

**Exercise 15: Replacing Motive Values**

The dataset creator noted that many entries in the `motive` column have very low counts, with some appearing only once, often in non-traditional series. To clean this column:

1) Make a copy of the `scoobydoo_merged` DataFrame and name it `scoobydoo_motives`. Use the new DataFrame for this exercise.
2) Identify the top five most common motives in the `motive` column.
3) Replace any value not in the top five with `Other`.
4) Check the value counts of the updated motive column and save the value counts as `Exercise15`.

In [30]:
### ENTER CODE HERE ###
scoobydoo_motives = scoobydoo_merged.copy()
top_five_motives = scoobydoo_motives['motive'].value_counts().head(5)
scoobydoo_motives['motive'] = scoobydoo_motives['motive'].apply(lambda x: x if x in top_five_motives else 'Other')
Exercise15 = scoobydoo_motives['motive'].value_counts()
Exercise15
scoobydoo_motives.shape

(603, 75)

**Code Check:** As a check, after replacing the appropriate values, you should have 188 `Other` entries.

In [31]:
### CODE CHECK ###
scoobydoo_motives['motive'].value_counts()

motive
Other               188
Competition         168
Theft               125
Treasure             54
Conquer              42
Natural Resource     26
Name: count, dtype: int64

<hr/>

**Exercise16: Binning IMDB Values**

Let's now practice binning the IMDB values. Your task is to put these values into the following categorical bins: `unknown`,`0-4.9`,`5-5.9`,`6-6.9`,`7-7.9`,`8-8.9`,`9-9.9`,`10`.

1) Make a copy of the `scoobydoo_merged` DataFrame and name it `scoobydoo_binning`. Use the new DataFrame for this exercise.
2) The category labels should be ordered and exactly match the above in the exact order.
3) Missing values should be replaced with the `unknown` category.

In [34]:
### ENTER CODE HERE ###
scoobydoo_binning = scoobydoo_merged.copy()
bins = [0, 4.9, 5.9, 6.9, 7.9, 8.9, 9.9, 10]
labels = ['0-4.9', '5-5.9', '6-6.9', '7-7.9', '8-8.9', '9-9.9','10']
scoobydoo_binning['imdb'] = pd.cut(scoobydoo_binning['imdb'], bins=bins, labels=labels)
scoobydoo_binning['imdb'] = scoobydoo_binning['imdb'].cat.add_categories('unknown').fillna('unknown')
new_order = ['unknown', '0-4.9', '5-5.9', '6-6.9', '7-7.9', '8-8.9', '9-9.9','10']
scoobydoo_binning['imdb'] = scoobydoo_binning['imdb'].cat.reorder_categories(new_order)
scoobydoo_binning.shape

(603, 75)

<hr/>

**Exercise17: Create Terrain Dummy Values**

Practice creating dummy variables for the `setting_terrain` column.

1) Make a copy of the `scoobydoo_merged` DataFrame and name it `scoobydoo_terrain`. Use the new DataFrame for this exercise.
2) Encode the `setting_terrain` column into dummy variables, ensuring to drop the first category and set the data type to integers. Use a prefix of `terrain` for the new columns (e.g., `terrain_Urban`, `terrain_Rural`, etc.)

In [37]:
### ENTER CODE HERE ###
scoobydoo_terrain = scoobydoo_merged.copy()
scoobydoo_terrain = pd.get_dummies(scoobydoo_terrain, columns=['setting_terrain'], drop_first=True, dtype=int, prefix='terrain')
scoobydoo_terrain.shape

(603, 88)

<hr/>

**Exercise18: Catchphrases Per Minute**

For those of you that have ever watched the show, I'm sure you've heard their catchphrases often: 'jeepers', 'jinkies', 'my_glasses', 'just_about_wrapped_up', 'zoinks', 'groovy', 'scooby_doo_where_are_you', 'rooby_rooby_roo'

Analyze which catchphrase is used the most per minute across all episodes.

1) Make a copy of the `scoobydoo_merged` DataFrame and name it `scoobydoo_catchphrase`. Use the new DataFrame for this exercise.
2) In this new DataFrame, calculate a per-minute value for each catchphrase by dividing the respective catchphrase column by the total minutes in the run_time column. Round the results to two decimal places and name the new columns with a `_per_minute` suffix (e.g., `jeepers_per_minute`, `jinkies_per_minute`, etc.).
3) Once the new columns are created, select only these columns, sum their values, and sort the results in descending order. Save the final output as a Series named `Exercise18`.  

In [38]:
### ENTER CODE HERE ###
scoobydoo_catchphrase = scoobydoo_merged.copy()
catchphrase_columns = ['jeepers', 'jinkies', 'my_glasses', 'just_about_wrapped_up', 'zoinks', 'groovy', 'scooby_doo_where_are_you', 'rooby_rooby_roo']
for col in catchphrase_columns:
    scoobydoo_catchphrase[f'{col}_per_minute'] = (scoobydoo_catchphrase[col] / scoobydoo_catchphrase['run_time']).round(2)
scoobydoo_catchphrase
Exercise18 = scoobydoo_catchphrase[[col + '_per_minute' for col in catchphrase_columns]].sum().sort_values(ascending=False)
Exercise18


zoinks_per_minute                      66.98
rooby_rooby_roo_per_minute             22.49
jinkies_per_minute                     19.14
jeepers_per_minute                      9.36
scooby_doo_where_are_you_per_minute     3.37
my_glasses_per_minute                   1.59
just_about_wrapped_up_per_minute        0.89
groovy_per_minute                       0.66
dtype: float64

**Code Check:** As a check, `zoinks_per_minute` should be 66.98.

In [43]:
### CODE CHECK ###
scoobydoo_catchphrase['zoinks_per_minute'].sum()
scoobydoo_catchphrase.shape

(603, 83)

<hr/>

**Exercise19: Working with Strings**

If you take a look at the `monster_name`, `monster_gender`, `monster_type`, `monster_subtype`, and the `monster_species` columns, you will notice that some columns have multiple values separated by commas when more than one monster is in the episode. Your goal is to separate these values so that each monster and its corresponding attributes are placed in their own new columns.

1) Make a copy of the `scoobydoo_merged` DataFrame and name it `scoobydoo_strings`. Use the new DataFrame for this exercise.
2) For each of the monster columns mentioned above, split the values into separate columns. Name the new columns using the pattern `<original_column_name>_part_#`, where # represents the sequence number of the monster (e.g., `monster_name_part_1`, `monster_name_part_2`, etc.). Ensure that the corresponding attributes (e.g., gender, type) follow the same pattern (e.g., `monster_gender_part_1`, `monster_gender_part_2`, etc.).
3) Make sure that you strip any leading or trailing spaces for the values.
    - Run the following code check to ensure the trailing space before "Disguised" is removed. You should see four occurrences of "Disguised": `scoobydoo_strings['monster_type_part_11'].value_counts()`

For example, let's say that an instance of `monster_name` contained three values: `Black Knight`, `Werewolf`, `Dracula`. `Black Knight` would end up in a new column called `monster_name_part_1`. `Werewolf` would end up in a new column called `monster_name_part_2`. `Dracula` would end up in a new column called `monster_name_part_3`. Their respective monster attributes would work similar such as `monster_gender_part_1` for the `Black Knight`'s gender, `monster_gender_part_2` for the `Werewolf`'s gender, `monster_gender_part_3` for `Dracula`'s gender, etc.

Hint: There are many ways to approach this problem. One approach is to use nested loops. The outer loop can iterate through the list of monster columns, and the inner loop can split the values and create the new columns with the appropriate suffixes.

In [41]:
### ENTER CODE HERE ###
scoobydoo_strings = scoobydoo_merged.copy()
monster_columns = ['monster_name', 'monster_gender', 'monster_type', 'monster_subtype', 'monster_species']
for col in monster_columns:
    parts = scoobydoo_strings[col].str.split(',', expand=True)
    for i in range(parts.shape[1]):
        scoobydoo_strings[f'{col}_part_{i+1}'] = parts[i].str.strip()
scoobydoo_strings
scoobydoo_strings['monster_type_part_11'].value_counts()

monster_type_part_11
Super-Villain    4
Disguised        4
Undead           3
Disfigured       1
Name: count, dtype: int64

<hr/>

**Exercise20: Pivot Tables**

Create a pivot table to analyze the average IMDB scores, broken down by network for the rows and format for the columns. Ensure that missing values are filled with `0` in the resulting pivot table, and include both row and column totals in your pivot table. Save the result as `Exercise20`.

In [42]:
### ENTER CODE HERE ###

Exercise20 = pd.pivot_table(scoobydoo_merged, values='imdb', index='network', columns='format', aggfunc='mean', fill_value=0, margins=True)

Exercise20

format                Crossover     Movie  Movie (Theatrical)  TV Series  \
network                                                                    
ABC                    6.800000  6.100000                 0.0   7.463725   
Adult Swim             8.100000  0.000000                 0.0   0.000000   
Boomerang              0.000000  0.000000                 0.0   7.572881   
CBS                    0.000000  0.000000                 0.0   7.902041   
Cartoon Network        7.633333  6.040000                 0.0   8.046053   
Syndication            0.000000  6.966667                 0.0   0.000000   
TBC                    0.000000  5.800000                 0.0   0.000000   
The CW                 9.600000  0.000000                 0.0   5.603846   
The WB                 0.000000  0.000000                 0.0   7.429268   
Warner Bros. Picture   0.000000  0.000000                 5.3   0.000000   
Warner Home Video      0.000000  6.518182                 0.0   6.900000   
All                    7.625000  6.467442                 5.3   7.516713   

format                TV Series (segmented)       All  
network                                                
ABC                                7.005143  7.166192  
Adult Swim                         0.000000  8.100000  
Boomerang                          0.000000  7.572881  
CBS                                0.000000  7.902041  
Cartoon Network                    0.000000  7.911905  
Syndication                        0.000000  6.966667  
TBC                                0.000000  5.800000  
The CW                             0.000000  5.751852  
The WB                             0.000000  7.429268  
Warner Bros. Picture               0.000000  5.300000  
Warner Home Video                  0.000000  6.576923  
All                                7.005143  7.277891

Congrats on completing the Scooby-Doo data assignment! You've unmasked data mysteries and earned your spot in the Scooby gang as a Numpy and Pandas pro. Keep solving those data mysteries!

In [98]:
### CODE CHECK ###
print(scoobydoo_merged.shape)

(603, 77)
